gobblin-utility/src/main/java/gobblin/util/executors/IteratorExecutor.java

```diff
@@ -59,29 +59,32 @@ public IteratorExecutor(Iterator<Callable<T>> runnableIterator, int numThreads,
    */
   public List<Future<T>> execute() throws InterruptedException {
 
-    if (this.executed) {
-      throw new RuntimeException(String.format("This %s has already been executed.", IteratorExecutor.class.getSimpleName()));
-    }
-
     List<Future<T>> futures = Lists.newArrayList();
-    int activeTasks = 0;
-    while (this.iterator.hasNext()) {
-      futures.add(this.completionService.submit(this.iterator.next()));
-      activeTasks++;
-      if (activeTasks == this.numThreads) {
+
+    try {
+      if (this.executed) {
+        throw new RuntimeException(String.format("This %s has already been executed.", IteratorExecutor.class.getSimpleName()));
+      }
+
+      int activeTasks = 0;
+      while (this.iterator.hasNext()) {
+        futures.add(this.completionService.submit(this.iterator.next()));
+        activeTasks++;
+        if (activeTasks == this.numThreads) {
+          this.completionService.take();
+          activeTasks--;
+        }
+      }
+      while (activeTasks > 0) {
         this.completionService.take();
         activeTasks--;
       }
-    }
-    while (activeTasks > 0) {
-      this.completionService.take();
-      activeTasks--;
-    }
 
-    this.completionService.poll();
-
-    ExecutorsUtils.shutdownExecutorService(this.executor, Optional.of(log), 10, TimeUnit.SECONDS);
-    this.executed = true;
+      this.completionService.poll();
```

Comment:
Is `poll` necessary here? `take` seems have taken care of all completion service?